charmhelpers/contrib/storage/linux/ceph.py

```diff
@@ -856,19 +856,35 @@ def _keyring_path(service):
     return KEYRING.format(service)
 
 
-def create_keyring(service, key):
-    """Create a new Ceph keyring containing key."""
+def add_key(service, key):
+    """
+    Add a key to a keyring.
+
+    Creates the keyring if it doesn't already exist.
+
+    Logs and returns if the key is already in the keyring.
+    """
     keyring = _keyring_path(service)
     if os.path.exists(keyring):
-        log('Ceph keyring exists at %s.' % keyring, level=WARNING)
-        return
+        with open(keyring) as ring:
+            if key not in ring.read():
+                log('Updating existing keyring %s.' % keyring, level=WARNING)
+            else:
+                log('Ceph keyring exists at %s and has not changed.' % keyring,
+                    level=WARNING)
+                return
 
     cmd = ['ceph-authtool', keyring, '--create-keyring',
            '--name=client.{}'.format(service), '--add-key={}'.format(key)]
     check_call(cmd)
     log('Created new ceph keyring at %s.' % keyring, level=DEBUG)
```

Comment:
This log message should probably be updated. It could be misleading if adding a key to an existing keyring.

Reply:
That log message only shows up if the keyring exists *and* the key is in the keyring. So the message in that case is accurate -- we'll perform a noop.